### PR TITLE
PLU-208: feat: lettersg integration part 1

### PR DESCRIPTION
**Pull request by m0nggh**
TODO:
- [x] Remove prod suffix for api key
- [x] Add extra monitor for letters action because the pdf download takes more than 10s?
- [x] Gate it to specific users?

## Problem
HSA and Sentosa case needs letters integration support on Plumber

## Solution
- Add [createLetter](https://guide.letters.gov.sg/developer-guide/upcoming-api-documentation) as a new action for staging, will test for prod once their API is out

## Tests
- Check that letter can be created on Plumber and the variables can be used
image

Note: Staging API key is granted on plumber@open.gov.sg, so you can access the key on 1password to test.
Letters generated with the Certificate of curiosity template won't show up on dashboard for now.

### CODE REVIEW

File: packages/backend/src/apps/lettersg/auth/is-still-verified.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/is-still-verified.ts]: nit: consider the simpler Typescript syntax `!authData?.apiKey` to improve readability

File: packages/backend/src/apps/lettersg/dynamic-data/get-template-ids.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/dynamic-data/get-template-ids.ts]: seems like we could simply `return { data: data.templates.map(...) }`, and drop the need for having an intermediate `templateIdsMap` variable

File: packages/backend/src/apps/lettersg/common/api.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/common/api.ts]: nit: might there be a more specific error type in the codebase that we could be using?

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts:141]:
```suggestion
        const [ firstError ] = fromZodError(error).details
```

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: might this wording be slightly better?
```suggestion
        'Specify each field name and value for your letter template.',
```

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/verify-api-key.ts]: should we be checking for `HTTP 401 Unauthorized` instead?

**Review by liangyuanruo — DISMISSED** [review]
Mostly nits, except for checking of authorization. We shouldn't be unpacking data fields to determine whether our API key was valid or not.

**Comment by liangyuanruo** [comment]
Have a think about logging also - is there any useful info that would help debug or recover from, if one of the steps were to break?

File: packages/backend/src/apps/lettersg/common/api.ts
Comment (jacksonOGP) [reply packages/backend/src/apps/lettersg/common/api.ts]: can we change the error message to more user friendly ones? "LetterSG API key is invalid" seems to be good enough instead of using the word "prefix".

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (jacksonOGP) [reply packages/backend/src/apps/lettersg/auth/verify-api-key.ts]: can consider changing the error message to "API key is invalid, please ensure you've copied the correct API key or contact support@plumber.gov.sg if you require assistance" 

Or

"API key is invalid, please ensure you've copied the correct API key"

Comment (m0nggh) [reply packages/backend/src/apps/lettersg/auth/verify-api-key.ts]: ok will do so, thanks!

File: packages/backend/src/apps/lettersg/common/api.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/common/api.ts]: for now, this is just a straightforward error message, so our codebase just throws a standard error

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/actions/create-letter/index.ts:141]: not sure which is more readable actually 😅

**Comment by m0nggh** [comment]
the data in the step is stored in `dataIn` and `dataOut` so I don't think there is a need to log more

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/verify-api-key.ts]: would it be a good idea to put in additional logging here for further investigation? if we throw an error with just this error message, I think we will lose the HTTP status and any informative error messages from LetterSG.

File: packages/backend/src/apps/lettersg/dynamic-data/get-template-ids.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/dynamic-data/get-template-ids.ts:34]: most likely the `catch` block here will be triggered by failure in `$.http.get('/v1/templates')` - should we add logging so that we can investigate why that might fail?

File: packages/backend/src/apps/lettersg/auth/is-still-verified.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/is-still-verified.ts]: any chance this code still works if we delete this line and update the return signature to `void`? Technically the return signature of this function is `Promise<true>`, as it never actually returns `false` anywhere

File: packages/backend/src/apps/lettersg/dynamic-data/get-template-fields.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/dynamic-data/get-template-fields.ts:39]: do we want to log this error?

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: To check whether this is a union of success and error types, as they usually don't come together.

For example:

```typescript
type LettersErrorData = {
  success: false
  message: string
  errors: Record<string, string>[]
}

type LettersSuccessData = {
  success: true
  message: string
}
```

Then TypeScript can perform type narrowing for you as long as you check the `success` flag

File: packages/backend/src/apps/lettersg/auth/verify-credentials.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/verify-credentials.ts]:
```suggestion
  if (!authData?.apiKey) {
```

File: packages/backend/src/apps/lettersg/auth/verify-credentials.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/verify-credentials.ts:25]:
```suggestion
  const labelSuffix = env === LetterSgEnvironment.Prod ? ' [PROD]' : ' [STAGING]'
  }
```

Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/auth/verify-credentials.ts]: To remove label suffixes for actual production release, as our users should not be seeing `[PROD]`

File: packages/backend/src/apps/lettersg/dynamic-data/get-template-fields.ts
Comment (liangyuanruo) [inline packages/backend/src/apps/lettersg/dynamic-data/get-template-fields.ts:19]:
```suggestion
      const { data, status } = await $.http.get(`/v1/templates/${templateId}`)
```

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/auth/verify-api-key.ts]: ok thks for pointing out, i realise that I should throw back the `err` here and log it under the graphql query instead

**Comment by liangyuanruo** [comment]
LGTM, feel free to ping on Slack for approval once part 2 is ready

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: this is a a specific error for only their validation of fields in the template, they will update us with a more specific error type, so I'm just gonna leave this out instead

File: packages/backend/src/graphql/queries/test-connection.ts
Comment (ogp-weeloong) [inline packages/backend/src/graphql/queries/test-connection.ts]: nit: should we add an `event` field to make it easier to filter? I typically do stuff like this to make it easier to search on DD.

https://github.com/opengovsg/plumber/blob/cd8ccd11d77841cdb215581b03dd2dd4870be68c/packages/backend/src/controllers/webhooks/handler.ts#L63

File: packages/backend/src/graphql/queries/test-connection.ts
Comment (ogp-weeloong) [inline packages/backend/src/graphql/queries/test-connection.ts]: Any reason we're nesting this inside `errorDetails`? Typically stepId, etc etc are at the root level of a log - e.g. 

https://github.com/opengovsg/plumber/blob/cd8ccd11d77841cdb215581b03dd2dd4870be68c/packages/backend/src/helpers/backoff.ts#L45-L50

aside: at some point we need to standardize how we log and add lints, otherwise its going to be harder and harder to search for stuff in DD

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: is this still TBC?

File: packages/backend/src/apps/lettersg/actions/create-letter/schema.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/actions/create-letter/schema.ts]: nit: can we also check for duplicates

nit: can we check for duplicates. e.g.

https://github.com/opengovsg/plumber/blob/cd8ccd11d77841cdb215581b03dd2dd4870be68c/packages/backend/src/apps/m365-excel/actions/update-table-row/schemas.ts#L20-L33

File: packages/backend/src/apps/lettersg/actions/create-letter/schema.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/actions/create-letter/schema.ts]: actually i think ok to remove nullish if the response is useless without it.

Thinking is that we don't want the pipe to silently succeed

File: packages/backend/src/apps/lettersg/auth/is-still-verified.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/auth/is-still-verified.ts]: nit: can add zod parsing.

https://github.com/opengovsg/plumber/blob/cd8ccd11d77841cdb215581b03dd2dd4870be68c/packages/backend/src/apps/m365-excel/common/auth-data.ts#L7-L12

File: packages/backend/src/apps/lettersg/auth/is-still-verified.ts
Comment (ogp-weeloong) [reply packages/backend/src/apps/lettersg/auth/is-still-verified.ts]: This is invoked by the `registerConnection` and `testConnection` mutations, both of which inspect the return value, so this can't be typed as void. But typing as `true` is fine.

We should actually return true if `verifyApiKey` succeeds, and false if it throws.

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/auth/verify-api-key.ts:8]: comment: maybe we should enhance StepError to make step number optional <.<

File: packages/backend/src/apps/lettersg/dynamic-data/types.ts
Comment (ogp-weeloong) [inline packages/backend/src/apps/lettersg/dynamic-data/types.ts:9]: nit: zod this in separate PR

**Review by ogp-weeloong — APPROVED** [review]
Generally LGTM, nice work + good tests

No need to address all comments now, but please do address these 3 before landing:
* Removing `[PROD]` label
* Move `stepId` etc out of `errorDetails` when logging test connection failures
* Remove `nullish` when zod is parsing LetterSG's response

File: packages/backend/src/graphql/queries/test-connection.ts
Comment (m0nggh) [reply packages/backend/src/graphql/queries/test-connection.ts]: thanks for the reco

File: packages/backend/src/apps/lettersg/actions/create-letter/index.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/actions/create-letter/index.ts]: yeapp will put a todo here to unblock

File: packages/backend/src/apps/lettersg/actions/create-letter/schema.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/actions/create-letter/schema.ts]: ok got it, thanks!

File: packages/backend/src/apps/lettersg/auth/verify-api-key.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/auth/verify-api-key.ts:8]: i think for now, `StepError` should be just for testing step but can look into this in the future!

File: packages/backend/src/apps/lettersg/auth/is-still-verified.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/auth/is-still-verified.ts]: roger sir

File: packages/backend/src/apps/lettersg/auth/verify-credentials.ts
Comment (m0nggh) [reply packages/backend/src/apps/lettersg/auth/verify-credentials.ts]: done!

**Pull request merged**
